### DISCUSSION TITLE
How to use auto incremented id's with MySQL when `IDENTITY` strategy is deprecated?

### DISCUSSION THREAD

**Discussion by ruudk** [doctrine/doctrine2#10739] [Support Questions] [answered]
We (and I think many older projects) heavily rely on auto incremented id's on MySQL. We use the `IDENTITY` strategy to automatically inject the last inserted ID into the entity.

Example:
```php
#[Entity]
class MyEntity
{
    #[Id]
    #[Column(name: 'id', type: 'integer')]
    #[GeneratedValue(strategy: 'IDENTITY')]
    private $id;
}
```

I noticed that since a recent update, this is now marked as deprecated. What is the suggested migration path when on MySQL?

It didn't became clear from the issues:
- https://github.com/doctrine/orm/pull/9833
- https://github.com/doctrine/orm/issues/8850
- https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/basic-mapping.html

**Comment by Fedik** [comment]
~For auto-incremented id use of `#[GeneratedValue]` should be enough, [by default it use `strategy: AUTO`.](https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/basic-mapping.html#identifiers-primary-keys), which will use IDENTITY for MySQL.~
I am a bit misunderstood the question.

**Comment by ruudk** [comment]
@greg0ire would you happen to know the answer to this? It seems you created the deprecation in https://github.com/doctrine/orm/commit/3f8430459cdad5293fb7cb7df83f1ad76bcb0b7f. Does it mean that `IDENTITY` (`AUTO`) on MySQL is deprecated?

**Reply by greg0ire** [reply]
Uuuuh does MySQL not support IDENTITY columns? What does `usesSequenceEmulatedIdentityColumns()` return for the platform you are using?

**Reply by ruudk** [reply]
`AbstractMySQLPlatform` does not overwrite `usesSequenceEmulatedIdentityColumns`, so therefore it delegates it to `Doctrine\DBAL\Platforms\AbstractPlatform::usesSequenceEmulatedIdentityColumns` that returns false and triggers the deprecation:
https://github.com/doctrine/dbal/blob/9a747d29e7e6b39509b8f1847e37a23a0163ea6a/src/Platforms/AbstractPlatform.php#L3943-L3953

Maybe the `AbstractPlatform::usesSequenceEmulatedIdentityColumns` should not trigger the deprecation?

**Reply by greg0ire** [reply]
If it returns `false`, you should not get a deprecation. And that's exactly what happens in `ClassMetadataFactory`, right?

**Reply by greg0ire** [reply]
Oh I get it now, you are getting a deprecation because `usesSequenceEmulatedIdentityColumns` is called. Would have been great to mention which deprecation you were having in your question, because as you can see, there are many of them.

**Reply by greg0ire** [reply, marked as answer]
So IDENTITY is not deprecated, what is really deprecated is calling `usesSequenceEmulatedIdentityColumns`. Do you confirm that's the deprecation you are getting?

**Reply by ruudk** [reply]
You're right, I was confused, sorry for that. 

It's indeed the deprecation coming from `AbstractPlatform::usesSequenceEmulatedIdentityColumns`. 

So the problem is that `ClassMetadataFactory` calls something that is deprecated, and should not call it.

**Reply by greg0ire** [reply]
Yes, and that's the case for ORM 3, but I don't see a way of avoiding that in ORM 2, unless maybe testing if the platform is an instance of this or that :grimacing:
